Fix and Improve sql logging coloration in `ActiveRecord::LogSubscriber`.

pboling opened the pull request:
- Fixes coloring for SQL statements generated with Ruby heredoc, which often have spaces preceding the initial SQL verb, like:

```
   sql = <<-EOS
      SELECT * FROM THINGS
        WHERE ID IN (
          SELECT ID FROM THINGS
        )
   EOS
```

^ Would be colored as MAGENTA in the existing code.
- Improves coloring for statements like:

```
    # Become WHITE
    SELECT * FROM (
        SELECT * FROM mytable FOR UPDATE
    ) ss WHERE col1 = 5;
    LOCK TABLE table_name IN ACCESS EXCLUSIVE MODE;

    # Becomes RED
    ROLLBACK
```
- Reinstates the coloration of the `payload[:name]` via new method:
  `colorize_payload_name`
  Instead of simple alternating colors, adds meaning:
  - `MAGENTA` for `"SQL"` or `blank?` payload names
  - `CYAN` for Model Load/Exists
- Make some `ActiveRecord::LogSubscriber` instance methods private for clarity:
  - `colorize_payload_name`
  - `sql_color`
  - `logger`
- Introduces specs for sql coloration.
- Introduces specs for payload name coloration.

GH #20885

pboling committed:
Improve sql logging coloration in `ActiveRecord::LogSubscriber`.

- Improves coloring for statements like:

    # Become WHITE
    SELECT * FROM (
        SELECT * FROM mytable FOR UPDATE
    ) ss WHERE col1 = 5;
    LOCK TABLE table_name IN ACCESS EXCLUSIVE MODE;

    # Becomes RED
    ROLLBACK

- Reinstates the coloration of the `payload[:name]`.
  Instead of simple alternating colors, adds meaning:
  - `MAGENTA` for `"SQL"` or `blank?` payload names
  - `CYAN` for Model Load/Exists

- Introduces specs for sql coloration.
- Introduces specs for payload name coloration.

GH#20885

egilburg commented on activerecord/lib/active_record/log_subscriber.rb:
```diff
@@ -55,12 +59,22 @@ def sql(event)
-        when /\s*\Adelete/i      then RED
-        when /transaction\s*\Z/i then CYAN
-        else MAGENTA
+        when /\A\s*rollback/mi then
```
don't need `then` keywords if using multi-line case statement.

pboling replied:
Thanks.  I am still used to writing code that is backwards compatible with Ruby vOLD. :P

pboling committed:
Remove extraneous `then` from case statement

- CR feedback from @egilburg

Additionally

- Move logic for colorizing the payload name into a separate method
- Make some `ActiveRecord::LogSubscriber` instance methods private for clarity:
  - `colorize_payload_name`
  - `sql_color`
  - `logger`
- Improve Changelog Documentation

GH #20885
